packages/cdk8s-plus/src/pod.ts

```diff
@@ -5,52 +5,173 @@ import * as cdk8s from 'cdk8s';
+export interface IPodTemplate extends IPodSpec {
 
   /**
-   * Provides access to the underlying spec.
-   *
-   * You can use this field to apply post instantiation mutations
-   * to the spec.
+   * Provides read/write access to the underlying pod metadata of the resource.
    */
-  public readonly spec: PodSpecDefinition;
+  readonly podMetadata: ApiObjectMetadataDefinition;
+}
 
-  constructor(scope: Construct, id: string, props: PodProps = {}) {
-    super(scope, id, props);
+/**
+ * Provides read/write capabilities ontop of a `PodSpecProps`.
+ */
+export class PodSpec implements IPodSpec {
 
-    this.spec = new PodSpecDefinition(props.spec);
+  public readonly restartPolicy?: RestartPolicy;
+  public readonly serviceAccount?: IServiceAccount;
 
-    this.apiObject = new k8s.Pod(this, 'Pod', {
-      metadata: props.metadata,
-      spec: cdk8s.Lazy.any({ produce: () => this.spec._toKube() }),
-    });
+  private readonly _containers: Container[];
+  private readonly _volumes: Volume[];
+
+  constructor(props: PodSpecProps = {}) {
+    this.restartPolicy = props.restartPolicy;
+    this.serviceAccount = props.serviceAccount;
+
+    this._containers = props.containers ?? [];
+    this._volumes = props.volumes ?? [];
+  }
+
+  public get containers(): Container[] {
+    return [ ...this._containers ];
+  }
+
+  public get volumes(): Volume[] {
+    return [ ...this._volumes ];
+  }
+
+  public addContainer(container: Container): void {
+    this._containers.push(container);
   }
+
+  public addVolume(volume: Volume): void {
+    this._volumes.push(volume);
+  }
+
+  /**
+   * @internal
+   */
+  public _toPodSpec(): k8s.PodSpec {
+
+    if (this.containers.length === 0) {
+      throw new Error('PodSpec must have at least 1 container');
+    }
+
+    const volumes: k8s.Volume[] = [];
+    const containers: k8s.Container[] = [];
+
+    for (const container of this.containers) {
+
+      // automatically add volume from the container mount
+      // to this pod so thats its available to the container.
+      for (const mount of container.mounts) {
+        volumes.push(mount.volume._toKube());
+      }
+
+      containers.push(container._toKube());
+    }
+
+    for (const volume of this._volumes) {
+      volumes.push(volume._toKube());
+    }
+
+    return {
+      restartPolicy: this.restartPolicy,
+      serviceAccountName: this.serviceAccount?.name,
+      containers: containers,
+      volumes: volumes,
+    };
+
+  }
+
 }
 
 /**
- * Properties for initialization of `PodSpec`.
+ * Provides read/write capabilities ontop of a `PodTemplateProps`.
+ */
+export class PodTemplate extends PodSpec implements IPodTemplate {
+
+  public readonly podMetadata: ApiObjectMetadataDefinition;
+
+  constructor(props: PodTemplateProps = {}) {
+    super(props);
+    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
+  }
+
+  /**
+   * @internal
+   */
+  public _toPodTemplateSpec(): k8s.PodTemplateSpec {
```

Comment:
This felt right, letting this class be responsible for creating the template spec used by `Job` and `Deployment`. It did mean I have to change the name if we want to preserve a typed return value from these methods.

We can also make this class not extend `PodSpec` but rather compose it and then we can use the normal `_toKube`, but that felt a bit redundant.

Reply:
Makes me think we need to rename all these toKube methods to their respective output type